rig/machine_control/scp_connection.py

```diff
@@ -138,7 +160,124 @@ def send_scp(self, buffer_size, x, y, p, cmd, arg1=0, arg2=0, arg3=0,
                 self.n_tries)
         )
 
-    def read(self, buffer_size, x, y, p, address, length_bytes):
+    def send_scp_burst(self, buffer_size, window_size,
+                       parameters_and_callbacks):
+        """Send a burst of SCP packets and call a callback for each returned
+        packet.
+
+        Parameters
+        ----------
+        buffer_size : int
+            Number of bytes held in an SCP buffer by SARK, determines how many
+            bytes will be expected in a socket.
+        window_size : int
+            Number of packets which can be awaiting replies from the SpiNNaker
+            board.
+        parameters_and_callbacks: iterable of :py:class:`.scpcall`
+            Iterable of :py:class:`.scpcall` elements.  These elements can
+            specify a callback which will be called with the returned packet.
+        """
+        # Non-blocking and then the following is a busy loop.
+        self.sock.setblocking(False)
+
+        # Calculate the receive length, this should be the smallest power of
+        # two greater than the required size
+        max_length = buffer_size + consts.SDP_HEADER_LENGTH
+        receive_length = 1 << 9  # 512 bytes is a reasonable minimum
+        while receive_length < max_length:
+            receive_length <<= 1
+
+        class TransmittedPacket(object):
+            """A packet which has been transmitted and still awaits a response.
+            """
+            __slots__ = ["callback", "packet", "expected_args", "n_tries",
+                         "time_sent"]
+
+            def __init__(self, callback, packet, expected_args):
+                self.callback = callback
+                self.packet = packet
+                self.expected_args = expected_args
+                self.n_tries = 1
+                self.time_sent = time.time()
+
+        queued_packets = True
+        outstanding_packets = {}
+
+        # While there are packets in the queue or packets for which we are
+        # still awaiting returns then continue to loop.
+        while queued_packets or outstanding_packets:
+            # If there are fewer outstanding packets than the window can take
+            # then transmit a packet and add it to the list of outstanding
+            # packets.
+            if len(outstanding_packets) < window_size:
```

Comment:
Oops...

```python
if len(outstanding_packets) < window_size and queued_packets:
```

I suspect an exception being raised for the last few times round this loop isn't good.